File: benchmarks-QuantumOptics.jl/multiplication_sparse_ket.jl
using QuantumOptics
using BenchmarkTools
include("benchmarkutils.jl")

srand(0)

basename = "multiplication_sparse_ket"

samples = 2
evals = 5
cutoffs = [100:100:1000;]
S = [0.1, 0.01, 0.001]
Nrand = 5

function setup(N, s)
    b = GenericBasis(N)
    op1 = SparseOperator(b, sprand(Complex128, N, N, s))
    psi = randstate(b)
    result = copy(psi)
    op1, psi, result
end

function f(op1, psi, result)
    operators.gemv!(Complex128(1., 0.), op1, psi, Complex128(0., 0.), result)
end

for s in S
    name = basename * "_" * replace(string(s), ".", "")
    println("Benchmarking: ", name)
    print("Cutoff: ")
    results = []
    for N in cutoffs
        print(N, " ")
        T = 0.
        for i=1:Nrand
            op1, psi, result = setup(N, s)
            T += @belapsed f($op1, $psi, $result) samples=samples evals=evals
        end
        push!(results, Dict("N"=>N, "t"=>T/Nrand))
    end
    println()
    benchmarkutils.save(name, results)
end

File: benchmarks-QuTiP/multiplication_sparse_ket.py
import scipy.sparse as sp
import qutip as qt
import benchmarkutils

basename = "multiplication_sparse_ket"

samples = 2
evals = 5
cutoffs = range(100, 1001, 100)
S = [0.1, 0.01, 0.001]
Nrand = 5


def setup(N, s):
    data = sp.rand(N, N, s, dtype=float) * 0.2j
    op1 = qt.Qobj(data)
    psi = qt.rand_ket(N).full().ravel()
    return op1, psi


def f(op1, psi):
    return op1 * psi


for s in S:
    name = basename + "_" + str(s).replace(".", "")
    print("Benchmarking:", name)
    print("Cutoff: ", end="", flush=True)
    results = []
    for N in cutoffs:
        print(N, "", end="", flush=True)
        T = 0.
        for i in range(Nrand):
            op1, psi = setup(N, s)
            T += benchmarkutils.run_benchmark(f, op1, psi,
                                              samples=samples, evals=evals)
        results.append({"N": N, "t": T / Nrand})
    print()
    benchmarkutils.save(name, results)
